### PR TITLE
Migrate version check

### LINKED ISSUE #3401: Migrations must refuse to run if it is unsafe
## Motivation

The zero-downtime upgrade may require that all running pulp processes, i.e. workers, pulp-content, and pulp-api, are all at plugin or pulpcore version X or newer.

The zero-downtime group will ensure the database layer is backwards compatible by having db changes being only additive, but at some point a breaking db change must occur. For example if you want to rename a column, first add a new column, use both columns for a while, and then eventually remove the old column. The removal part of this example would be the breaking change.

## Proposal

The goal is to ensure migrations are always safe, and stop users from applying unsafe migrations.

* Breaking change migrations specify which versions of which components "must be" running before this change can be allowed
* Such migrations start by querying the DB for the current status (via heartbeat-msgs) of everything running in the system/cluster
* If anything is running at lower-than-required, the migration is halted
* In this situation, the "old" components are identified, and what-they-need-to-be-running is identified, and the admin is expected to upgrade/shut down any pods w/ old-code
* At that point, the upgrade can be restarted

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #3401

### CODE REVIEW

File: pulpcore/migrations.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        self.version = version
+        self.hints = hints or {}
+        self.elidable = True
+
+    def state_forwards(self, app_label, state):
+        pass
+
+    def database_forwards(self, app_label, schema_editor, from_state, to_state):
+        from_state.clear_delayed_apps_cache()
+        if router.allow_migrate(schema_editor.connection.alias, app_label, **self.hints):
```
Comment (mdellweg) [inline pulpcore/migrations.py:23]: Reminder: What is this for? Is it safe to skip this check?

**Comment by mdellweg** [comment]
Sample output:
```
RuntimeError: Incompatible versions detected (gem >= 0.4.0 needed):
  - 'gem'='0.3.0.dev' with api server '1996@e87410f347fa'
  - 'gem'='0.3.0.dev' with api server '1997@e87410f347fa'
  - 'gem'='0.3.0.dev' with content server '1998@e87410f347fa'
  - 'gem'='0.3.0.dev' with content server '2002@e87410f347fa'
  - 'gem'='0.3.0.dev' with pulp worker '703@e87410f347fa'
  - 'gem'='0.3.0.dev' with pulp worker '652@e87410f347fa'
Please shutdown or upgrade the outdated components before you continue the migration.
```

File: pulpcore/migrations.py (hunk 1 above)
Comment (mdellweg) [reply pulpcore/migrations.py]: I could not find a good explanation and decided skipping the test is probably not desired.

File: pulpcore/plugin/migrations.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from pulpcore.migrations import RequirePlugin  # noqa: F401
```
Comment (gerrod3) [inline pulpcore/plugin/migrations.py:1]: Am I missing something, isn't the migration called `RequireVersion`, not `RequirePlugin`?

Comment (mdellweg) [reply pulpcore/plugin/migrations.py:1]: Why is flake8 not complaining?

Comment (mdellweg) [reply pulpcore/plugin/migrations.py:1]: But the name would be up for discussion. ;)

Comment (gerrod3) [reply pulpcore/plugin/migrations.py]: I'm good with the name.

File: pulpcore/migrations.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+from packaging.version import parse as parse_version
+
+from django.conf import settings
+from django.utils import timezone
+from django.db.migrations.operations.base import Operation
+
+
+class RequireVersion(Operation):
+    reversible = True
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/migrations.py:9]: Well, it claims that migrating down is ok.
We do not check whether to new workers are running when migrating down. Neither do we check that you do not spin up old workers again after migrating up. There's only so much you can do...

**Review by gerrod3 — APPROVED** [review]

**Review by ggainey — APPROVED** [review]

**Pull request merged**